# Actividad - Proyecto práctico

> La actividad se desarrollará en grupos pre-definidos de 2-3 alumnos. Se debe indicar los nombres en orden alfabético (de apellidos). Recordad que esta actividad se corresponde con un 30% de la nota final de la asignatura. Se debe entregar entregar el trabajo en la presente notebook.

*   Alumno 1: Granizo, Mateo
*   Alumno 2: Maiolo, Pablo
*   Alumno 3: Miglino, Diego
  
https://github.com/dmiglino/SpaceInvaders_ReinforcementLearning/tree/entrega


---
## **PARTE 1** - Instalación y requisitos previos

> Las prácticas han sido preparadas para poder realizarse en el entorno de trabajo de Google Colab. Sin embargo, esta plataforma presenta ciertas incompatibilidades a la hora de visualizar la renderización en gym. Por ello, para obtener estas visualizaciones, se deberá trasladar el entorno de trabajo a local. Por ello, el presente dosier presenta instrucciones para poder trabajar en ambos entornos. Siga los siguientes pasos para un correcto funcionamiento:
1.   **LOCAL:** Preparar el enviroment, siguiendo las intrucciones detalladas en la sección *1.1.Preparar enviroment*.
2.  **AMBOS:** Modificar las variables "mount" y "drive_mount" a la carpeta de trabajo en drive en el caso de estar en Colab, y ejecturar la celda *1.2.Localizar entorno de trabajo*.
3. **COLAB:** se deberá ejecutar las celdas correspondientes al montaje de la carpeta de trabajo en Drive. Esta corresponde a la sección *1.3.Montar carpeta de datos local*.
4.  **AMBOS:** Instalar las librerías necesarias, siguiendo la sección *1.4.Instalar librerías necesarias*.


---
### 1.1. Preparar enviroment (solo local)



> Para preparar el entorno de trabajo en local, se han seguido los siguientes pasos:
1. En Windows, puede ser necesario instalar las C++ Build Tools. Para ello, siga los siguientes pasos: https://towardsdatascience.com/how-to-install-openai-gym-in-a-windows-environment-338969e24d30.
2. Instalar Anaconda
3. Siguiendo el código que se presenta comentado en la próxima celda: Crear un enviroment, cambiar la ruta de trabajo, e instalar librerías básicas.


```
conda create --name miar_rl python=3.8
conda activate miar_rl
cd "PATH_TO_FOLDER"
conda install git
pip install jupyter
```


4. Abrir la notebook con *jupyter-notebook*.



```
jupyter-notebook
```


---
### 1.2. Localizar entorno de trabajo: Google colab o local

In [1]:
# ATENCIÓN!! Modificar ruta relativa a la práctica si es distinta (drive_root)
mount='/content/gdrive'
drive_root = mount + "/My Drive/08_MIAR/actividades/TP_Grupal"
mount='./'

try:
  from google.colab import drive
  IN_COLAB=True
except:
  IN_COLAB=False
print(IN_COLAB)

False


---
### 1.3. Montar carpeta de datos local (solo Colab)

In [2]:
# Switch to the directory on the Google Drive that you want to use
import os
if IN_COLAB:
  print("We're running Colab")

  if IN_COLAB:
    # Mount the Google Drive at mount
    print("Colab: mounting Google drive on ", mount)

    drive.mount(mount)

    # Create drive_root if it doesn't exist
    create_drive_root = True
    if create_drive_root:
      print("\nColab: making sure ", drive_root, " exists.")
      os.makedirs(drive_root, exist_ok=True)

    # Change to the directory
    print("\nColab: Changing directory to ", drive_root)
    %cd $drive_root
# Verify we're in the correct working directory
%pwd
print("Archivos en el directorio: ")
print(os.listdir())

Archivos en el directorio: 
['.git', '.ipynb_checkpoints', 'anaconda_projects', 'apr_g9_dqn_SpaceInvaders-v0_4M_log.json', 'apr_g9_dqn_SpaceInvaders-v0_5M_GPU_Local_log.json', 'apr_g9_dqn_SpaceInvaders-v0_6M_log.json', 'apr_g9_dqn_SpaceInvaders-v0_log.json', 'apr_g9_dqn_SpaceInvaders-v0_weights_1000.h5f.data-00000-of-00001', 'apr_g9_dqn_SpaceInvaders-v0_weights_1000.h5f.index', 'apr_g9_dqn_SpaceInvaders-v0_weights_10000.h5f.data-00000-of-00001', 'apr_g9_dqn_SpaceInvaders-v0_weights_10000.h5f.index', 'apr_g9_dqn_SpaceInvaders-v0_weights_1000000_4M.h5f.data-00000-of-00001', 'apr_g9_dqn_SpaceInvaders-v0_weights_1000000_4M.h5f.index', 'apr_g9_dqn_SpaceInvaders-v0_weights_1000000_5M_GPU_Local.h5f.data-00000-of-00001', 'apr_g9_dqn_SpaceInvaders-v0_weights_1000000_5M_GPU_Local.h5f.index', 'apr_g9_dqn_SpaceInvaders-v0_weights_100000_5M_GPU_Local.h5f.data-00000-of-00001', 'apr_g9_dqn_SpaceInvaders-v0_weights_100000_5M_GPU_Local.h5f.index', 'apr_g9_dqn_SpaceInvaders-v0_weights_1000_5M_GPU_Local.

---
### 1.4. Instalar librerías necesarias

In [3]:
# ejecutar solo la primera vez..

if IN_COLAB:
  %pip install gym==0.17.3
  %pip install git+https://github.com/Kojoley/atari-py.git
  %pip install keras-rl2==1.0.5
  %pip install tensorflow==2.12
else:
  %pip install gym==0.17.3
  %pip install git+https://github.com/Kojoley/atari-py.git
  %pip install pyglet==1.5.0
  %pip install h5py==3.1.0
  %pip install Pillow==9.5.0
  %pip install keras-rl2==1.0.5
  %pip install Keras==2.2.4
  %pip install tensorflow==2.5.3
  %pip install torch==2.0.1
  %pip install agents==1.4.0
  %pip install matplotlib==3.3.4

Note: you may need to restart the kernel to use updated packages.
  Cloning https://github.com/Kojoley/atari-py.git to c:\users\dmigl\appdata\local\temp\pip-req-build-r02cc9bg
  Resolved https://github.com/Kojoley/atari-py.git to commit 86a1e05c0a95e9e6233c3a413521fdb34ca8a089
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
Note: you may need to restart the kernel to use updated packages.


  Running command git clone --filter=blob:none --quiet https://github.com/Kojoley/atari-py.git 'C:\Users\dmigl\AppData\Local\Temp\pip-req-build-r02cc9bg'


Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
  Using cached numpy-1.19.5-cp38-cp38-win_amd64.whl.metadata (2.0 kB)
Using cached numpy-1.19.5-cp38-cp38-win_amd64.whl (13.3 MB)
  Attempting uninstall: numpy
    Found existing installation: numpy 1.24.4
    Uninstalling numpy-1.24.4:
      Successfully uninstalled numpy-1.24.4
Note: you may need to restart the kernel to use updated packages.


  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.


Note: you may need to restart the kernel to use updated packages.


Note: you may need to restart the kernel to use updated packages.


---
## **PARTE 2**. Enunciado

Consideraciones a tener en cuenta:

- El entorno sobre el que trabajaremos será _SpaceInvaders-v0_ y el algoritmo que usaremos será _DQN_.

- Para nuestro ejercicio, el requisito mínimo será alcanzado cuando el agente consiga una **media de recompensa por encima de 20 puntos en modo test**. Por ello, esta media de la recompensa se calculará a partir del código de test en la última celda del notebook.

Este proyecto práctico consta de tres partes:

1.   Implementar la red neuronal que se usará en la solución
2.   Implementar las distintas piezas de la solución DQN
3.   Justificar la respuesta en relación a los resultados obtenidos

**Rúbrica**: Se valorará la originalidad en la solución aportada, así como la capacidad de discutir los resultados de forma detallada. El requisito mínimo servirá para aprobar la actividad, bajo premisa de que la discusión del resultado sera apropiada.

IMPORTANTE:

* Si no se consigue una puntuación óptima, responder sobre la mejor puntuación obtenida.
* Para entrenamientos largos, recordad que podéis usar checkpoints de vuestros modelos para retomar los entrenamientos. En este caso, recordad cambiar los parámetros adecuadamente (sobre todo los relacionados con el proceso de exploración).
* Se deberá entregar unicamente el notebook y los pesos del mejor modelo en un fichero .zip, de forma organizada.
* Cada alumno deberá de subir la solución de forma individual.

---
## **PARTE 3**. Desarrollo y preguntas

#### Importar librerías

In [3]:
import tensorflow as tf

print("Versión de TensorFlow:", tf.__version__)
gpus = tf.config.experimental.list_physical_devices('GPU')

if gpus:
    for gpu in gpus:
        tf.config.experimental.set_memory_growth(gpu, True)
    print("GPU detectada:", gpus)
else:
    print("No se detectó GPU")

Versión de TensorFlow: 2.5.3
GPU detectada: [PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]


In [4]:
from __future__ import division

from PIL import Image
import numpy as np
import gym

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Activation, Flatten, Convolution2D, Permute, Dropout
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.initializers import HeNormal
from tensorflow.keras.optimizers.schedules import ExponentialDecay
import tensorflow.keras.backend as K

from rl.agents.dqn import DQNAgent
from rl.policy import LinearAnnealedPolicy, BoltzmannQPolicy, EpsGreedyQPolicy
from rl.memory import SequentialMemory
from rl.core import Processor
from rl.callbacks import FileLogger, ModelIntervalCheckpoint

#### Configuración base

In [5]:
INPUT_SHAPE = (84, 84)
WINDOW_LENGTH = 4

env_name = 'SpaceInvaders-v0'
env = gym.make(env_name)

np.random.seed(123)
env.seed(123)
nb_actions = env.action_space.n

print("Numero de acciones disponibles: " + str(nb_actions) + " -> " + str(env.unwrapped.get_action_meanings()))
print("Formato de las observaciones: " + str(env.observation_space))

Numero de acciones disponibles: 6 -> ['NOOP', 'FIRE', 'RIGHT', 'LEFT', 'RIGHTFIRE', 'LEFTFIRE']
Formato de las observaciones: Box(0, 255, (210, 160, 3), uint8)


In [6]:
class AtariProcessor(Processor):
    def process_observation(self, observation):
        assert observation.ndim == 3  # (height, width, channel)
        # removemos la parte superior (score) y inferior (vida del jugador)
        cropped = observation[20:200, 8:152]
        img = Image.fromarray(cropped)
        img = img.resize(INPUT_SHAPE, Image.LANCZOS).convert('L')
        processed_observation = np.array(img)
        assert processed_observation.shape == INPUT_SHAPE
        return processed_observation.astype('uint8')

    def process_state_batch(self, batch):
        processed_batch = batch.astype('float32') / 255.
        return processed_batch

    def process_reward(self, reward):
        return np.clip(reward, -1., 1.)

1. Implementación de la red neuronal

In [7]:
# modelo mas complejo - usar este
input_shape = (WINDOW_LENGTH,) + INPUT_SHAPE
model = Sequential()
print(K.image_data_format())
print(input_shape)

if K.image_data_format() == 'channels_last':
    # (width, height, channels)
    model.add(Permute((2, 3, 1), input_shape=input_shape))
elif K.image_data_format() == 'channels_first':
    # (channels, width, height)
    model.add(Permute((1, 2, 3), input_shape=input_shape))
else:
    raise RuntimeError('Unknown image_dim_ordering.')

# Capas convolucionales mejoradas con inicialización HeNormal
model.add(Convolution2D(32, (8, 8), strides=(4, 4), 
                       activation='relu', kernel_initializer=HeNormal()))
 # model.add(BatchNormalization())

model.add(Convolution2D(64, (4, 4), strides=(2, 2), 
                       activation='relu', kernel_initializer=HeNormal()))
 # model.add(BatchNormalization())

model.add(Convolution2D(64, (3, 3), strides=(1, 1), 
                       activation='relu', kernel_initializer=HeNormal()))
 # model.add(BatchNormalization())

model.add(Flatten())

# Capas densas con dropout y batch normalization
model.add(Dense(512, activation='relu', kernel_initializer=HeNormal()))
 # model.add(BatchNormalization())
model.add(Dropout(0.3))

model.add(Dense(256, activation='relu', kernel_initializer=HeNormal()))
 # model.add(BatchNormalization())
model.add(Dropout(0.2))

# Capa de salida
model.add(Dense(nb_actions, activation='linear', kernel_initializer=HeNormal()))

print(model.summary())

channels_last
(4, 84, 84)
Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
permute (Permute)            (None, 84, 84, 4)         0         
_________________________________________________________________
conv2d (Conv2D)              (None, 20, 20, 32)        8224      
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 9, 9, 64)          32832     
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 7, 7, 64)          36928     
_________________________________________________________________
flatten (Flatten)            (None, 3136)              0         
_________________________________________________________________
dense (Dense)                (None, 512)               1606144   
_________________________________________________________________
dropout (Dropout)            (

2. Implementación de la solución DQN

In [8]:
from rl.callbacks import Callback

class HyperparameterMonitorCallback(Callback):
    """Callback para monitorear el estado de la memoria de los actuales valores de los hiperparametros"""
    
    def __init__(self, interval=5000):
        self.interval = interval
        self.steps_since_log = 0
        
    def on_step_end(self, step, logs=None):
        self.steps_since_log += 1
        if self.steps_since_log >= self.interval:
            self.steps_since_log = 0
            
            print(f"\n>>> Memoria e Hiperparametros:")
            
            # Estado de la memoria/buffer
            memory = self.model.memory
            print(f"    [MEMORY] Memory Size: {len(memory)}")
            print(f"    [MEMORY] Elite Fraction: {memory.elite_fraction}")
            
            # Valores de los hiperparametros
            print(f"    [TRAIN] Batch Size: {dqn.batch_size}")
            print(f"    [TRAIN] Step actual: {self.model.step}")
            
            policy = getattr(self.model, 'policy', None)
            if isinstance(policy, LinearAnnealedPolicy):
                eps = policy.get_current_value()
                print(f"    [POLICY] epsilon actual: {eps:.4f}")
            optimizer = self.model.model.optimizer
            lr_schedule = optimizer.learning_rate

            if isinstance(lr_schedule, ExponentialDecay):
                # Usamos el número de paso actual para evaluar el schedule
                current_lr = float(lr_schedule(self.model.step).numpy())
            else:
                # Puede ser una variable o valor fijo
                current_lr = float(K.get_value(lr_schedule))
            print(f"    [LR] Learning rate actual: {current_lr:.6f}")
            

    def on_episode_end(self, episode, logs=None):
        """Mostrar stats al final del episodio"""
        pass

In [9]:
from rl.callbacks import Callback

class BatchSizeSchedulerCallback(Callback):
    def __init__(self, total_steps, initial_bs=32, schedule=None):
        self.total_steps = total_steps
        self.initial_bs = initial_bs
        self.schedule = schedule if schedule else [(1/3, 2), (2/3, 2)]
        self.applied = [False] * len(self.schedule)

    def on_train_begin(self, logs=None):
        self.model.batch_size = self.initial_bs
        print(f"[BATCH] Entrenamiento iniciado con batch_size={self.initial_bs}")

    def on_step_end(self, step, logs=None):
        current_step = self.model.step  # paso global del agente

        for i, (threshold_frac, multiplier) in enumerate(self.schedule):
            if not self.applied[i] and current_step >= int(self.total_steps * threshold_frac):
                new_bs = self.model.batch_size * multiplier
                self.model.batch_size = new_bs
                self.applied[i] = True
                print(f"[BATCH] Step {current_step}: batch_size actualizado a {new_bs}")


In [17]:
from rl.callbacks import Callback
import numpy as np
import math

class AdaptiveEliteTuningAndSaveBestModelCallback(Callback):
    def __init__(self, 
                 env,
                 filename='best_model.h5f',
                 test_episodes=5,
                 warmup_steps=1000,
                 monitor_interval=1000,
                 improve_threshold=1.0,
                 decay_threshold=0.1,
                 decay_factor=0.7,
                 grow_factor=0.7,
                 min_elite=0.1,
                 max_elite=0.9):
        
        self.env = env
        self.filename = filename
        self.test_episodes = test_episodes
        self.warmup_steps = warmup_steps
        self.monitor_interval = monitor_interval
        self.improve_threshold = improve_threshold
        self.decay_threshold = decay_threshold
        self.decay_factor = decay_factor
        self.grow_factor = grow_factor
        self.min_elite = min_elite
        self.max_elite = max_elite
        self.best_avg_reward = -np.inf
        self.steps_since_last_eval = 0

    def on_step_end(self, step, logs=None):
        self.steps_since_last_eval += 1
        
        if self.steps_since_last_eval >= self.monitor_interval:
            self.steps_since_last_eval = 0
            saved_step = self.model.step

            # Evaluación
            history = self.model.test(
                self.env,
                nb_episodes=self.test_episodes,
                visualize=False,
                verbose=0
            )
            
            avg_reward = np.mean(history.history['episode_reward'])
            max_reward = np.max(history.history['episode_reward'])
            delta = avg_reward - self.best_avg_reward

            # Restaurar contador del modelo (importante para evitar errores internos)
            self.model.step = saved_step
            self.model.training = True

            print(f"\n>>> Step {saved_step} -> Avg={avg_reward:.2f}, Max={max_reward:.2f} -> {history.history['episode_reward']} -> Mejor Actual: {self.best_avg_reward:.2f}")

            # Guardar pesos si mejora
            if avg_reward > self.best_avg_reward:
                self.best_avg_reward = avg_reward
                self.model.save_weights(self.filename, overwrite=True)
                print(f">>> Nuevo Máximo: {avg_reward:.2f} -> Guardado en {self.filename}")

            # Ajuste de elite_fraction si la memoria lo permite
            if self.model.step >= self.warmup_steps and not math.isinf(delta):
                mem = getattr(self.model, 'memory', None)
                current_elite = getattr(mem, 'elite_fraction', None)
    
                if current_elite is not None:
                    print(f"[AdaptiveElite] Δ={delta:.2f}, elite_fraction={current_elite:.2f}")
                    if delta > self.improve_threshold:
                        new_elite = min(current_elite * self.grow_factor, self.max_elite)
                        print(f"[AdaptiveElite] Mejora → Aumento elite_fraction a {new_elite:.2f}")
                        mem.elite_fraction = new_elite
                    elif delta < self.decay_threshold:
                        new_elite = max(current_elite * self.decay_factor, self.min_elite)
                        print(f"[AdaptiveElite] Estancamiento → Reduzco elite_fraction a {new_elite:.2f}")
                        mem.elite_fraction = new_elite


In [10]:
import numpy as np
import copy
import random
from collections import deque
from rl.memory import sample_batch_indexes, zeroed_observation

class ElitistSequentialMemory(SequentialMemory):
    def __init__(self, limit, window_length=1, elite_fraction=0.5, batch_size=128, reward_threshold=0.0, **kwargs):
        super().__init__(limit=limit, window_length=window_length, **kwargs)
        self.elite_fraction = elite_fraction
        self.batch_size = batch_size
        self.reward_threshold = reward_threshold
        self.limit = limit
        self.window_length = window_length

        self.elite_fraction = elite_fraction  # porcentaje del batch que será elite
        self.reward_threshold = reward_threshold  # criterio de recompensa mínima para ser elite
        
        self.actions = []
        self.rewards = []
        self.terminals = []
        self.observations = []

    def append(self, observation, action, reward, terminal, training=True):
        super().append(observation, action, reward, terminal, training=training)
        if not training:
            return

        if reward < 0:
            print(f"\n[MEMORY] NEGATIVA {reward:.2f}")

        if len(self.observations) < self.limit:
            self.observations.append(observation)
            self.actions.append(action)
            self.rewards.append(reward)
            self.terminals.append(terminal)
            # print(f"\n[MEMORY] Agregada experiencia nueva con reward {reward:.2f}")
        else:
            # Buscar el índice de la peor experiencia actual
            min_reward = min(self.rewards)
            if reward > min_reward:
                idx = self.rewards.index(min_reward)
                #if reward > 0:
                #    print(f"\n[MEMORY] Reemplazada exp (old: {min_reward:.2f}) -> nueva: {reward:.2f}")
                self.observations[idx] = observation
                self.actions[idx] = action
                self.rewards[idx] = reward
                self.terminals[idx] = terminal

    def sample(self, batch_size, batch_idxs=None):
        all_experiences = super().sample(batch_size * 4)
    
        # Definimos una recompensa mínima para considerar élite (podés ajustar)
        rewards = [e.reward for e in all_experiences]
        threshold = np.percentile(rewards, 100 * (1 - self.elite_fraction))  # por ejemplo, top 20%
    
        elite = [e for e in all_experiences if e.reward >= threshold]
        elite_ids = set(id(e) for e in elite)
    
        elite_count = min(len(elite), int(self.elite_fraction * batch_size))
        selected_elite = elite[:elite_count]
    
        remaining = batch_size - len(selected_elite)
        others = [e for e in all_experiences if id(e) not in elite_ids]
    
        if len(others) >= remaining:
            final_others = others[:remaining]
        else:
            # Si no hay suficientes "otros", completamos duplicando de los mejores
            needed = remaining - len(others)
            if len(others) > 0:
                final_others = others + random.choices(others, k=needed)
            else:
                final_others = random.choices(selected_elite, k=remaining)
        
        final_batch = selected_elite + final_others
        #print(f"\n[MEMORY] Se muestrea un batch de {batch_size} con {len(selected_elite)} exp de elite y {len(final_others)} restantes.")
        assert len(final_batch) == batch_size
        return final_batch

    @property
    def nb_entries(self):
        return len(self.observations)

    def get_config(self):
        config = super().get_config()
        config['limit'] = self.limit
        return config
    
    def __len__(self):
        return len(self.observations)


In [18]:
from tensorflow.keras.optimizers.schedules import ExponentialDecay

# memoria para almacenar la experiencia del agente
# memory = SequentialMemory(limit=10000, window_length=WINDOW_LENGTH)
# memory = EpisodeParameterMemory(limit=100000, window_length=WINDOW_LENGTH)
# memory = PrioritizedMemory(limit=100000, alpha=.6, start_beta=.4, end_beta=1., steps_annealed=100000, window_length=WINDOW_LENGTH)

# Memoria con priorizacion de mejores episodios 
# memory = PrioritizedExperienceReplay(
#     limit=3000,           # Tamaño de memoria
#     alpha=0.75,              # Priorización: 0.6 es un buen balance
#     beta_start=0.4,         # Importance sampling inicial
#     beta_frames=2500,     # Frames para annealing (75% del entrenamiento)
#     epsilon=1e-6,           # Evitar prioridades cero
#     window_length=WINDOW_LENGTH,
#    ignore_episode_boundaries=False
# )
# print(f"Prioritized Experience Replay configurado: alpha={memory.alpha}, beta_start={memory.beta_start}")
memory = ElitistSequentialMemory(
    limit=2000000,
    window_length=4,
    elite_fraction=0.7,     # Usar 70% de experiencias con buen reward
    reward_threshold=0.5    # Solo se considera “elite” si el reward > 0.5
)
print(f"Elitist Sequential Memory configurada: elite_fraction={memory.elite_fraction}, batch_size={memory.batch_size}, reward_threshold={memory.reward_threshold}")

# processor
processor = AtariProcessor()

# policy que el agente va a seguir
#policy = BoltzmannQPolicy()
policy = LinearAnnealedPolicy(EpsGreedyQPolicy(), attr='eps',
                              value_max=1.00, value_min=0.08, 
                              value_test=.05, nb_steps=900000)
print(f"Estrategia configurada: value_max={policy.value_max}, value_min={policy.value_min}, nb_steps={policy.nb_steps}")

# definicion del agente
dqn = DQNAgent(model=model, nb_actions=nb_actions, policy=policy,
               memory=memory, processor=processor,
               nb_steps_warmup=50000, gamma=.994, batch_size=32,
               target_model_update=10000, train_interval=4, delta_clip=1.0,
               enable_double_dqn=True, enable_dueling_network=False)
print(f"Agente configurado: nb_steps_warmup={dqn.nb_steps_warmup}, gamma={dqn.gamma}, target_model_update={dqn.target_model_update}")

# compilacion del agente
initial_learning_rate = 0.0002  # Learning rate más conservador
lr_schedule = ExponentialDecay(
    initial_learning_rate,
    decay_steps=100000,
    decay_rate=0.96,
    staircase=True
)
dqn.compile(Adam(learning_rate=lr_schedule), metrics=['mae'])
print(f"Learning Rate configurado: initial_lr={initial_learning_rate}, decay_rate={lr_schedule.decay_rate}, decay_steps={lr_schedule.decay_steps}")

Elitist Sequential Memory configurada: elite_fraction=0.7, batch_size=128, reward_threshold=0.5
Estrategia configurada: value_max=1.0, value_min=0.08, nb_steps=900000
Agente configurado: nb_steps_warmup=50000, gamma=0.994, target_model_update=10000
Learning Rate configurado: initial_lr=0.0002, decay_rate=0.96, decay_steps=900000


In [19]:
nombre_modelo = 'Agente_1M_GPU_Local'

# entrenamiento del agent
#weights_filename = 'apr_g9_dqn_{}_weights.h5f'.format(env_name)
tf.debugging.set_log_device_placement(True)
checkpoint_weights_filename = nombre_modelo + '/apr_g9_dqn_' + env_name + '_weights_{step}_1M_GPU_Local.h5f'
log_filename = nombre_modelo + '/apr_g9_dqn_{}_1M_GPU_Local_log.json'.format(env_name)
total_steps = 1000000

callbacks = [
    ModelIntervalCheckpoint(checkpoint_weights_filename, interval=total_steps/100),
    FileLogger(log_filename, interval=total_steps/100),
    #SaveBestRewardCallback(env, filename='best_model_weights_6M.h5f', test_episodes=5, test_interval=500),
    #AdaptiveEliteFractionCallback(env, monitor_interval=100, improve_threshold=1.0, decay_threshold=0.1, decay_factor=0.7, grow_factor=0.7, min_elite=0.2, max_elite=0.8),
    AdaptiveEliteTuningAndSaveBestModelCallback(env, filename=nombre_modelo + '/best_model_weights_1M_GPU_Local.h5f', test_episodes=5, 
                                                warmup_steps=dqn.nb_steps_warmup, monitor_interval=5000, 
                                                improve_threshold=1.0, decay_threshold=0.1, decay_factor=0.9, 
                                                grow_factor=1.1, min_elite=0.15, max_elite=0.85),
    BatchSizeSchedulerCallback(total_steps=total_steps, initial_bs=32, schedule=[(1/3, 2), (3/4, 2)]),
    HyperparameterMonitorCallback(interval=total_steps/100)
]
entrenamiento = dqn.fit(env, callbacks=callbacks, nb_steps=total_steps, log_interval=2500, visualize=False, verbose=1)

# se graban los pesos finales luego de finalizar el entrenamiento
#dqn.save_weights('apr_g9_dqn_{}_weights.h5f'.format(env_name), overwrite=True)

[BATCH] Entrenamiento iniciado con batch_size=32
Training for 1000000 steps ...
Interval 1 (0 steps performed)
2500/2500 [==============================] - 8s 3ms/step - reward: 0.0172
2 episodes - episode_reward: 11.500 [8.000, 15.000] - ale.lives: 2.333

Interval 2 (2500 steps performed)
2487/2500 [============================>.] - ETA: 0s - reward: 0.0125
>>> -inf

>>> Step 4999 -> Avg=11.80, Max=12.00 -> [12.0, 12.0, 12.0, 11.0, 12.0] -> Mejor Actual: -inf
>>> Nuevo Máximo: 11.80 -> Guardado en best_model_weights_1M_GPU_Local.h5f
2500/2500 [==============================] - 21s 9ms/step - reward: 0.0124
3 episodes - episode_reward: 12.667 [5.000, 22.000] - ale.lives: 2.385

Interval 3 (5000 steps performed)
2500/2500 [==============================] - 9s 3ms/step - reward: 0.0128
5 episodes - episode_reward: 8.200 [3.000, 13.000] - ale.lives: 1.968

Interval 4 (7500 steps performed)
2484/2500 [============================>.] - ETA: 0s - reward: 0.0153
>>> 11.8

>>> Step 9999 -> Avg

In [ ]:
from gym.wrappers import Monitor
import numpy as np

# cargar mejor modelo
dqn.load_weights(nombre_modelo + '/best_model_weights_1M_GPU_Local.h5f')

# test normal + guardar recompensas
history = dqn.test(env, nb_episodes=100, visualize=False)
episode_rewards = history.history['episode_reward']
np.save(nombre_modelo + '/test_rewards.npy', episode_rewards)

# Calcular métricas
promedio = np.mean(episode_rewards)
desviacion = np.std(episode_rewards)
mejor_reward = np.max(episode_rewards)
mejor_epi = np.argmax(episode_rewards) + 1

# Mostrar los resultados
print(f"\nPromedio de recompensas: {promedio:.2f}")
print(f"Desviación estándar: {desviacion:.2f}")
print(f"Mejor Episodio {mejor_epi} con Recompensa = {mejor_reward}")

# Guardar la mejor recompensa en un archivo
with open(nombre_modelo + '/mejor_ejecucion.txt', 'w') as f:
    f.write(f"Mejor episodio: {mejor_epi} con Recompensa: {mejor_reward}\n")
    f.write(f"Promedio: {promedio:.2f}, Desviación estándar: {desviacion:.2f}\n")

# correr episodio con video
# video_env = Monitor(gym.make(env_name), './video', force=True)
# dqn.test(video_env, nb_episodes=1, visualize=False)
# video_env.close()

Testing for 100 episodes ...
Episode 1: reward: 27.000, steps: 1163
Episode 2: reward: 14.000, steps: 816
Episode 3: reward: 19.000, steps: 1064
Episode 4: reward: 16.000, steps: 940
Episode 5: reward: 21.000, steps: 1149
Episode 6: reward: 21.000, steps: 925
Episode 7: reward: 17.000, steps: 851
Episode 8: reward: 22.000, steps: 937
Episode 9: reward: 19.000, steps: 864
Episode 10: reward: 11.000, steps: 676
Episode 11: reward: 19.000, steps: 1063
Episode 12: reward: 14.000, steps: 817
Episode 13: reward: 24.000, steps: 1106
Episode 14: reward: 22.000, steps: 1073
Episode 15: reward: 24.000, steps: 1169
Episode 16: reward: 24.000, steps: 1478
Episode 17: reward: 16.000, steps: 802
Episode 18: reward: 16.000, steps: 699
Episode 19: reward: 15.000, steps: 875
Episode 20: reward: 20.000, steps: 797
Episode 21: reward: 23.000, steps: 1259
Episode 22: reward: 15.000, steps: 818
Episode 23: reward: 18.000, steps: 878
Episode 24: reward: 16.000, steps: 634
Episode 25: reward: 17.000, steps: 

In [ ]:
import matplotlib.pyplot as plt
rewards = np.load(nombre_modelo + '/test_rewards.npy')

plt.figure(figsize=(14, 10))
plt.plot(rewards)
plt.title("Recompensa por episodio")
plt.xlabel("Episodio")
plt.ylabel("Recompensa")
plt.grid()
plt.savefig(nombre_modelo + '/grafico_recompensas.png')
plt.show()

In [ ]:
plt.figure(figsize=(14, 10))
plt.hist(rewards, bins=range(int(min(rewards)), int(max(rewards)) + 2), edgecolor='black')
plt.title('Histograma de Recompensas')
plt.xlabel('Recompensa por Episodio')
plt.ylabel('Frecuencia')
plt.grid(True)
plt.show()

3. Justificación de los parámetros seleccionados y de los resultados obtenidos

Luego de muchas pruebas y experimentos, llegamos a este agente que consiguió un promedio de 20 puntos, con un máximo de 32. Otros agentes previos llegaron a 40 y 41 puntos, pero con un promedio de recompensas menor entre 15 y 16 puntos. Por lo que valoramos más el promedio de los episodios por sobre un resultado individual que pueda ser superior. Asimismo, el actual modelo obtuvo una desviación estándar de X, mientras que otros modelos con mejor puntaje máximo mostraron mayor varianza, lo que los hace menos confiables.

A lo largo de estas semanas fuimos haciendo cambios y agregando complejidad al modelo. 
Lo más relevante posiblemente sea pasar de una SequentialMemory que provee keras-rl a otros tipos de memoria. En primer lugar se desarrolló una PrioritizedMemory, donde las experiencias se priorizaron para luego utilizar las mejores durante el muestreo. Esto no tenía mucho sentido, ya que las recompensas son discretas, es decir, una acción puede no dar recompensa, o si lo hace, da 1, no hay valores intermedios y la priorización carecía de sentido. Además, el agente también aprende de experiencias de acciones que no dieron recompensas. 
Por lo dicho, implementamos una memoria basada en la idea de elitismo, donde a la hora de tomar un muestreo de experiencias lo hacemos con un porcentaje de experiencias positivas, para favorecer la explotación, y el resto de experiencias aleatorias hasta completar el batch_size. Cuando más bajo sea dicho porcentaje, más se favorece la exploración para poder salir de óptimos locales.
La idea es que este porcentaje que mide el nivel de elitismo sea dinámico, por eso mediante un callback estamos evaluando si el modelo esta estancado o no, en caso de estarlo se va disminuyendo su valor, o aumentando en caso contrario. 
El actual modelo comienza con un valor del 60% de elitismo, y se ve como el mismo fue disminuyendo hasta llegar al mínimo de 15%.
Además, la memoria tiene una capacidad del doble de los steps del entrenamiento, ya que vimos que se llena por encima del valor de los steps. 

Respecto al batch_size, también es dinámico. La idea fue ir incrementandolo a medida que avanzaba el entrenamiento para tener resultados más estables hacia el final. Se comienza con un batch size de 32, que es el valor estándar, y mediante un callback (BatchSizeSchedulerCallback) se incremente al doble cuando llegamos al 33% del entrenamiento, y otra vez al doble al llegar al 75% del mismo. 

Para la Policy se usó un LinearAnnealedPolicy con el valor del epsilon que va decayendo desde 1 hasta 0.08 a lo largo del 90% del total de steps, en este caso, 900.000.

El agente DQN se compiló con un optimizador Adam con learning rate dinámico usando ExponentialDecay, el cual comienza en 0.0002 y cada 100.000 pasos se multiplica por 0.96, haciendo que disminuya su valor.

Este último agente se entrenó con 1M de pasos, lo que le alcanzó para superar en promedio de recompensas a otros agentes entrenados en 4, 5 y 6 millones de steps, pero con menor complejidad. Mediante el AdaptiveEliteTuningAndSaveBestModelCallback se evalúa el modelo cada 5000 pasos para quedarnos con los mejores pesos si al evaluarlo en 5 episodios, el promedio supera al máximo encontrado hasta ese momento.

Por último, como ayuda al modelo, en el AtariProcessor estamos haciendo un crop de las observaciones para eliminar los bordes laterales, y la información extra como el puntaje obtenido y la cantidad de vidas. También se agregó inicialización HeNormal a cada capa de la red neuronal así como un Dropout de 0.3 a las capas Dense, para ayudar a estabilizar el entrenamiento junto al batch_size elevado.

Además, cada 10.000 steps, monitoreamos el estado de la memoria y el valor de estos hiperparametros dinámicos con otro callback (HyperparameterMonitorCallback).

El diseño progresivo del agente priorizó la estabilidad y adaptabilidad. A pesar de entrenar con menos pasos que otros agentes previos, alcanzó un desempeño más confiable, demostrando que la calidad de las experiencias y la dinámica de los hiperparámetros son más relevantes que la cantidad de iteraciones.

El mejor modelo durante el entrenamiento se encontró cerca de la mitad del mismo, en el step 455.000, si bien luego se encontraron otros modelos cercanos a este, ninguno logró superar su promedio de recompensas, lo cual indica que podria implementarse un mecanismo para frenar el entrenamiento si pasa una cantidad determinada de episodios sin mejoras. 

---
*   Alumno 1: Granizo, Mateo
*   Alumno 2: Maiolo, Pablo
*   Alumno 3: Miglino, Diego
  
https://github.com/dmiglino/SpaceInvaders_ReinforcementLearning/tree/entrega